### ISSUE TITLE
The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?

### ISSUE THREAD

**Issue by logici** [canonical/microk8s#1916]
After joining cluster, and executed `microk8s kubectl get node `:

 `The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?`

ENV:
host : mac

- multipass ubuntu:20.04 * 2 (one: microk8s-vm-0 the other one: microk8s-vm-1)
- microk8s 1.19/stable

OPERATION:

1. multipass shell microk8s-vm-0 and `sudo snap install microk8s --classic --channel=1.19/stable`
2. 

```
sudo usermod -a -G microk8s $USER
sudo chown -f -R $USER ~/.kube
```

3. `microk8s status` :

```
microk8s is running
high-availability: no
  datastore master nodes: 127.0.0.1:19001
  datastore standby nodes: none
addons:
  enabled:
    ha-cluster           # Configure high availability on the current node
  disabled:
    ambassador           # Ambassador API Gateway and Ingress
    cilium               # SDN, fast with full network policy
    dashboard            # The Kubernetes dashboard
    dns                  # CoreDNS
    fluentd              # Elasticsearch-Fluentd-Kibana logging and monitoring
    gpu                  # Automatic enablement of Nvidia CUDA
    helm                 # Helm 2 - the package manager for Kubernetes
    helm3                # Helm 3 - Kubernetes package manager
    host-access          # Allow Pods connecting to Host services smoothly
    ingress              # Ingress controller for external access
    istio                # Core Istio service mesh services
    jaeger               # Kubernetes Jaeger operator with its simple config
    knative              # The Knative framework on Kubernetes.
    kubeflow             # Kubeflow for easy ML deployments
    linkerd              # Linkerd is a service mesh for Kubernetes and other frameworks
    metallb              # Loadbalancer for your Kubernetes cluster
    metrics-server       # K8s Metrics Server for API access to service metrics
    multus               # Multus CNI enables attaching multiple network interfaces to pods
    prometheus           # Prometheus operator for monitoring and logging
    rbac                 # Role-Based Access Control for authorisation
    registry             # Private image registry exposed on localhost:32000
    storage              # Storage class; allocates storage from host directory
```

4.  `microk8s insepct` :

```
Inspecting Certificates
Inspecting services
  Service snap.microk8s.daemon-cluster-agent is running
  Service snap.microk8s.daemon-containerd is running
  Service snap.microk8s.daemon-apiserver is running
  Service snap.microk8s.daemon-apiserver-kicker is running
  Service snap.microk8s.daemon-control-plane-kicker is running
  Service snap.microk8s.daemon-proxy is running
  Service snap.microk8s.daemon-kubelet is running
  Service snap.microk8s.daemon-scheduler is running
  Service snap.microk8s.daemon-controller-manager is running
  Copy service arguments to the final report tarball
Inspecting AppArmor configuration
Gathering system information
  Copy processes list to the final report tarball
  Copy snap list to the final report tarball
  Copy VM name (or none) to the final report tarball
  Copy disk usage information to the final report tarball
  Copy memory usage information to the final report tarball
  Copy server uptime to the final report tarball
  Copy current linux distribution to the final report tarball
  Copy openSSL information to the final report tarball
  Copy network configuration to the final report tarball
Inspecting kubernetes cluster
  Inspect kubernetes cluster
Inspecting juju
  Inspect Juju
Inspecting kubeflow
  Inspect Kubeflow

Building the report tarball
  Report tarball is at /var/snap/microk8s/1856/inspection-report-20210120_111529.tar.gz
```

5. All the nodes are fine. `microk8s kubectl get all --all-namespaces`

```
NAMESPACE     NAME                                          READY   STATUS    RESTARTS   AGE
kube-system   pod/calico-node-vsrjz                         1/1     Running   0          17m
kube-system   pod/calico-kube-controllers-847c8c99d-rgt4j   1/1     Running   0          17m

NAMESPACE   NAME                 TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
default     service/kubernetes   ClusterIP   10.152.183.1   <none>        443/TCP   18m

NAMESPACE     NAME                         DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
kube-system   daemonset.apps/calico-node   1         1         1       1            1           kubernetes.io/os=linux   17m

NAMESPACE     NAME                                      READY   UP-TO-DATE   AVAILABLE   AGE
kube-system   deployment.apps/calico-kube-controllers   1/1     1            1           17m

NAMESPACE     NAME                                                DESIRED   CURRENT   READY   AGE
kube-system   replicaset.apps/calico-kube-controllers-847c8c99d   1         1         1       17m
```

6. Microk8s-vm-0 executed `microk8s add-node` .

```
From the node you wish to join to this cluster, run the following:
microk8s join 192.168.xx.xx:25000/d67362036a00d3d44a0040a34e2e4f9d

If the node you are adding is not reachable through the default interface you can use one of the following:
 microk8s join 192.168.xx.xx:25000/d67362036a00d3d44a0040a34e2e4f9d
 microk8s join 10.1.245.0:25000/d67362036a00d3d44a0040a34e2e4f9d
```

microk8s-vm-1 and microk8s-vm-0 can ping each other. microk8s-vm-1 execute 

```
ubuntu@microk8s-vm-1:~$ microk8s join 192.168.xx.xx:25000/d67362036a00d3d44a0040a34e2e4f9d
Contacting cluster at 192.168.xx.xx
Waiting for this node to finish joining the cluster. .. .. .. .. .. .. .. .. .. ..  
```

7. execute result:

- microk8s status: `microk8s is not running. Use microk8s inspect for a deeper inspection.`

- microk8s inspect:

  ```
  Inspecting Certificates
  Inspecting services
    Service snap.microk8s.daemon-cluster-agent is running
    Service snap.microk8s.daemon-containerd is running
    Service snap.microk8s.daemon-apiserver is running
    Service snap.microk8s.daemon-apiserver-kicker is running
    Service snap.microk8s.daemon-control-plane-kicker is running
    Service snap.microk8s.daemon-proxy is running
    Service snap.microk8s.daemon-kubelet is running
    Service snap.microk8s.daemon-scheduler is running
    Service snap.microk8s.daemon-controller-manager is running
    Copy service arguments to the final report tarball
  Inspecting AppArmor configuration
  Gathering system information
    Copy processes list to the final report tarball
    Copy snap list to the final report tarball
    Copy VM name (or none) to the final report tarball
    Copy disk usage information to the final report tarball
    Copy memory usage information to the final report tarball
    Copy server uptime to the final report tarball
    Copy current linux distribution to the final report tarball
    Copy openSSL information to the final report tarball
    Copy network configuration to the final report tarball
  Inspecting kubernetes cluster
    Inspect kubernetes cluster
  Inspecting juju
    Inspect Juju
  Inspecting kubeflow
    Inspect Kubeflow
  
  Building the report tarball
    Report tarball is at /var/snap/microk8s/1856/inspection-report-20210120_112705.tar.gz
  ```

- microk8s kubectl get node :

   `The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?`

- `cluster.yaml` included two IP of `microk8s-vm-0` and `microk8s-vm-1`.

8. I'm sorry for that I dont attach some logs ,because it happened last week and I donot find it.

9. the same as the previous phenomenon， you can recurrent it by joining itself.

10. ```
    ubuntu@microk8s-vm-1:~$ microk8s add-node
    From the node you wish to join to this cluster, run the following:
    microk8s join 192.168.xx.xx:25000/6f7b943c08560c366b7fc7ceaa66043a
    
    If the node you are adding is not reachable through the default interface you can use one of the following:
     microk8s join 192.168.xx.xx:25000/6f7b943c08560c366b7fc7ceaa66043a
     microk8s join 10.1.245.0:25000/6f7b943c08560c366b7fc7ceaa66043a
     
     
     
     ubuntu@microk8s-vm-1:~$ microk8s join 192.168.xx.xx:25000/6f7b943c08560c366b7fc7ceaa66043a
    Contacting cluster at 192.168.xx.xx
    Waiting for this node to finish joining the cluster. .. .. .. .. .. .. .. .. .. ..  
    ```

11. specially, `cluster.yaml` is different from the previous situation.

```
cat /var/snap/microk8s/1856/var/kubernetes/backend/cluster.yaml 
- Address: 192.168.64.4:19001
  ID: 0
  Role: 0

cat /var/snap/microk8s/1856/var/kubernetes/backend/info.yaml 
Address: 192.168.64.4:19001
ID: 13153073102629576865
Role: 0

```

12. maybe they have the same phenomenon, but the reason lead to the phenomenon is different.Once I recurrent the first situation, I update it in time.
13. At present, I resolve it by removing  all the files except `cluster.crt` and `cluster.key` in the `/var/snap/microk8s/1856/var/kubernetes/backend/`.And restart microk8s.

**Comment by balchua** [comment]
Can you check if you can do `ping microk8s-vm-1` from `microk8s-vm-0` and vice versa?

If not can you try adding them to your `/etc/hosts`?

**Comment by balchua** [comment]
Oops sorry i missed rhat one.  The `cluster-agent` logs may reveal something.  One can force remove a node though.  I think its `microk8s remove-node node_name --force`

**Comment by logici** [comment]
I'm so sorry. I have tested for two days so far, but it could not recurrent again. The same phenomenon that was `The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?` happened then, I had tried many methods. I only recovered it by reinstalling `microk8s` or deleting some files in my 13th step `operation`.

**Comment by adthonb** [comment]
I get the same problem when joining a node. My two nodes in the Singapore and I want to join one node in London. All have public IP but London's node can't join.

**Mine inspect here**
[inspection-report-20210122_095459.tar.gz](https://github.com/ubuntu/microk8s/files/5855028/inspection-report-20210122_095459.tar.gz)

**Comment by tobiolalusi** [comment]
I get the same error as well. I could sometimes run `microk8s kubectl get no` and get all the expected nodes and sometimes I get the error `The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?` instead.

**Comment by balchua** [comment]
@tobiolalusi its likely that the apiserver is restarting.  Sometimes Network interface changes can cause the apiserver to be restarted by the apiservice-kicker component.

**Comment by tobiolalusi** [comment]
@balasu It must have been so. I restarted my machines again and everything seems to have been working fine so far so good.

**Comment by keltroth** [comment]
I my case, it was right after trying to add a external node to my cluster... I had to start from scratch a new microk8s cluster. Hope I helped...

**Comment by balchua** [comment]
@adthonb does your cluster already have plenty of workloads running?

**Comment by mfpnca** [comment]
For me, this start each time I reboot a node

**Comment by jamesgeddes** [comment]
I've got a clean install running on a RPi 3 and I'm getting this. Any nudges in the right direction would be much appreciated.

**Comment by balchua** [comment]
@jamesgeddes in my experience in the past using rpi3 with k8s is sluggish due to its 1GB memory.  
But i've not tried it with the recent release of MicroK8s 1.21 where one of the features is to combine k8s components into 1 binary.  If you want to try it `sudo snap install microk8s --channel 1.21/stable --classic`

**Comment by jamesgeddes** [comment]
> @jamesgeddes my experience in the past using rpi3 with k8s is sluggish due to its 1GB memory.  
> But i've not tried it with the recent release of MicroK8s 1.21 where one of the features is to combine k8s components into 1 binary.  If you want to try it `sudo snap install microk8s --channel 1.21/stable --classic`

I do already have it installed. I am not worried about the speed, I am just a bit stumped about how to get past the "have you selected the right port" thing.

**Comment by balchua** [comment]
@jamesgeddes 
What does `journalctl -u snap.microk8s.daemon-kubelite.service -n 1000` say?

**Comment by jamesgeddes** [comment]
@balasu 

> -- Logs begin at Thu 2020-09-24 19:27:12 UTC, end at Mon 2021-04-12 09:25:09 UTC. --
> -- No entries --

**Comment by balchua** [comment]
@jamesgeddes strange 1.21/stable uses kubelite.  Maybe i gave the wrong service name. Maybe 
`snap.microk8s.daemon-kubelite` is enuf.

**Comment by jamesgeddes** [comment]
That's running now. Just to be clear, micro k8s does run on the RPi 3 b+, right? I appreciate that it might be slow, but I'm only using it for my personal playground right now.

**Comment by jamesgeddes** [comment]
It's still running...

**Comment by jamesgeddes** [comment]
@balchua This is still running, has it hung do you think or is it just super slow?

**Comment by balchua** [comment]
Couldn't  say for certain, but Im guessing its still running but just super slow. 
The inspect command will try to get all the logs from MicroK8s systemd services, get kubernetes resources, grab network related items like iptables.  And then make a tarball out of it.

**Comment by jamesgeddes** [comment]
That has been running for over 7 hours now, so I'm going to guess it has died. I've restarted in a screen session and will come back to it tomorrow.

**Comment by Turiok** [comment]
@jamesgeddes  have you seen this doc : https://microk8s.io/docs/troubleshooting ?
For my case I have a RPI4 and RPI3. On RPI4, After renaming my hostname and the iptable. It works and after the installation the node has a load average to 1.70.
But on my RPI 3, with the same OS configuration, the load average is to 7 or 9. microk8s inspect say is running but not microk8s status. I don't know why. I see the same line for my rpi3 from this issue and not on my rpi4

**Comment by jamesgeddes** [comment]
@balchua `microk8s inspect` finally finished, had to leave it over night.

[inspection-report-20210413_103927.tar.gz](https://github.com/ubuntu/microk8s/files/6303181/inspection-report-20210413_103927.tar.gz)

@Turiok what did you name your boards?

**Comment by balchua** [comment]
Thanks @jamesgeddes i think the apiserver is getting oom killed by the kernel as shown from dmesg

```
[32449.984451] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/system.slice/snap.microk8s.daemon-apiserver.service,task=kube-apiserver,pid=6889,uid=0
[32449.984604] Out of memory: Killed process 6889 (kube-apiserver) total-vm:2010980kB, anon-rss:422668kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:1316kB oom_score_adj:0
```
You may need to enable swap, but it is not recommended by kubernetes tho.  For the purpose of getting it up maybe fine.  Im still skeptic it is usable. 🙁. You can still give it a try tho.

**Comment by jamesgeddes** [comment]
So do I need to buy RPi 4 boards or should it work on the RPi 3?

**Comment by Turiok** [comment]
I updated microk8s (1.21) on my rpi3 (named rpi-pink) and my rpi4 (named rpi-blue). I have the same problem now.
The rpis are on arch OS. I installed libselinux with the last microk8s version. If not systemctl services doesn't start.
I installed ifconfig. Without it, the log network/ifconfig didn't display right log.

I think it's a problem from iptables.  But I have already make the action from the documentation to accept the forwarding.
[inspection-report-20210413_134256.tar.gz](https://github.com/ubuntu/microk8s/files/6303446/inspection-report-20210413_134256.tar.gz)

**Comment by balchua** [comment]
> So do I need to buy RPi 4 boards or should it work on the RPi 3?

If it is up to me i will probably get the rpi4 8GB board to run with the control plane.  And i will use the rpi3 as a worker node.  This is not the typical  MicroK8s setup.  I recently wrote some tutorial on how to do this with MicroK8s.  But its going to be a little bit more involved.
https://discuss.kubernetes.io/t/adding-a-worker-node-only-with-microk8s/14801

**Comment by jamesgeddes** [comment]
Argh the [tutorial](https://ubuntu.com/tutorials/how-to-kubernetes-cluster-on-raspberry-pi) made it seem so simple! Seemed to suggest I could use Pi2 or Pi3 boards :cry:

**Comment by balchua** [comment]
@jamesgeddes today i tried setting up a single node with `1` CPU with `1GB` RAM on DigitalOcean cloud.  It is a `amd64` arch VM.
First i have to turn swap on, then install microk8s.  Wait for it to be ready, and then enable `rbac`, `dns` and `ingress` addons.
Current usage
![image](https://user-images.githubusercontent.com/28750768/115100456-51f30780-9f6f-11eb-9e89-45b7b277cb39.png)

```shell
microk8s status
microk8s is running
high-availability: no
  datastore master nodes: 127.0.0.1:19001
  datastore standby nodes: none
addons:
  enabled:
    dns                  # CoreDNS
    ha-cluster           # Configure high availability on the current node
    ingress              # Ingress controller for external access
    rbac                 # Role-Based Access Control for authorisation
  disabled:
    ambassador           # Ambassador API Gateway and Ingress
    cilium               # SDN, fast with full network policy
    dashboard            # The Kubernetes dashboard
    fluentd              # Elasticsearch-Fluentd-Kibana logging and monitoring
    gpu                  # Automatic enablement of Nvidia CUDA
    helm                 # Helm 2 - the package manager for Kubernetes
    helm3                # Helm 3 - Kubernetes package manager
    host-access          # Allow Pods connecting to Host services smoothly
    istio                # Core Istio service mesh services
    jaeger               # Kubernetes Jaeger operator with its simple config
    keda                 # Kubernetes-based Event Driven Autoscaling
    knative              # The Knative framework on Kubernetes.
    kubeflow             # Kubeflow for easy ML deployments
    linkerd              # Linkerd is a service mesh for Kubernetes and other frameworks
    metallb              # Loadbalancer for your Kubernetes cluster
    metrics-server       # K8s Metrics Server for API access to service metrics
    multus               # Multus CNI enables attaching multiple network interfaces to pods
    openebs              # OpenEBS is the open-source storage solution for Kubernetes
    openfaas             # openfaas serverless framework
    portainer            # Portainer UI for your Kubernetes cluster
    prometheus           # Prometheus operator for monitoring and logging
    registry             # Private image registry exposed on localhost:32000
    storage              # Storage class; allocates storage from host directory
    traefik              # traefik Ingress controller for external access
root@n1:~# microk8s kubectl get all -A
NAMESPACE     NAME                                          READY   STATUS    RESTARTS   AGE
ingress       pod/nginx-ingress-microk8s-controller-p6sgb   1/1     Running   0          31m
kube-system   pod/coredns-7f9c69c78c-25jgl                  1/1     Running   0          34m
kube-system   pod/calico-kube-controllers-f7868dd95-zgmpw   1/1     Running   0          43m
kube-system   pod/calico-node-5lst8                         1/1     Running   0          43m

NAMESPACE     NAME                 TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                  AGE
default       service/kubernetes   ClusterIP   10.152.183.1    <none>        443/TCP                  44m
kube-system   service/kube-dns     ClusterIP   10.152.183.10   <none>        53/UDP,53/TCP,9153/TCP   34m

NAMESPACE     NAME                                               DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
kube-system   daemonset.apps/calico-node                         1         1         1       1            1           kubernetes.io/os=linux   44m
ingress       daemonset.apps/nginx-ingress-microk8s-controller   1         1         1       1            1           <none>                   31m

NAMESPACE     NAME                                      READY   UP-TO-DATE   AVAILABLE   AGE
kube-system   deployment.apps/calico-kube-controllers   1/1     1            1           44m
kube-system   deployment.apps/coredns                   1/1     1            1           34m

NAMESPACE     NAME                                                DESIRED   CURRENT   READY   AGE
kube-system   replicaset.apps/calico-kube-controllers-f7868dd95   1         1         1       43m
kube-system   replicaset.apps/coredns-7f9c69c78c                  1         1         1       34m
root@n1:~# 
```
Your mileage may vary, since rpi arm64 devices are slower than those cloud VMs.  Still not a great experience though.

**Comment by jamesgeddes** [comment]
Thanks @balchua 

Seems like microk8s is not entirely compatible with the RPi3 so I gave up and installed portainer. Perhaps someone should update the Ubuntu.com instructions to note that one must use the RPi4. It's kinda frustrating as I have 2 x RPi3 and 4 x RPi4 boards just siting here, and I wasted time following the instructions unaware it was never going to work.

I will try again when I have a master that is a bit more capable. Thanks everyone for your help though.

**Comment by tix4soccer** [comment]
> Thanks @balchua
> 
> Seems like microk8s is not entirely compatible with the RPi3 so I gave up and installed portainer. Perhaps someone should update the Ubuntu.com instructions to note that one must use the RPi4. It's kinda frustrating as I have 2 x RPi3 and 4 x RPi4 boards just siting here, and I wasted time following the instructions unaware it was never going to work.
> 
> I will try again when I have a master that is a bit more capable. Thanks everyone for your help though.

I wouldn't bother with RPi4. It is not specific to earlier versions of the pi.
It also happens there.
I have 3 RPi4s.
All running same version of ubuntu (20.04) and all on microk8s ( 1.19 ).
One of them has the error mentioned here.

**Comment by SuperJMN** [comment]
> Thanks @balchua
> 
> Seems like microk8s is not entirely compatible with the RPi3 so I gave up and installed portainer. Perhaps someone should update the Ubuntu.com instructions to note that one must use the RPi4. It's kinda frustrating as I have 2 x RPi3 and 4 x RPi4 boards just siting here, and I wasted time following the instructions unaware it was never going to work.
> 
> I will try again when I have a master that is a bit more capable. Thanks everyone for your help though.

I'm also having problems with Raspberry Pi 3. No matter what I try, issues all around. I hope somebody can sort them out and make MicroK8s work in the Pi 3.

**Comment by Boes-man** [comment]
Have the same error on macOS. kubectl commands became slow. mulitpass commands also started erroring with list failed: ssh connection failed: 'Timeout connecting..' A multipass stop and then start (restarted errored) of the vm seemed to work, but after a few minutes things stopped working again. Had (did multipass delete) a few apps and services running in it, so think its resource related. Just concerning that multipass commands stopped working, looks like multipass (and microk8s) could become orphaned/unusable due to a microk8s vm oom...?

**Comment by odonzyk** [comment]
check your /etc/ hosts file there should be these lines

127.0.0.1 localhost 127.0.1.1 name you servers

**Comment by TheLazyGeekGuy** [comment]
Had the same issue, it was microk8s process that was down !

`
$> microk8s.status
microk8s is not running. Use microk8s inspect for a deeper inspection.

$> microk8s start
Started.

$> microk8s status
microk8s is running
high-availability: no
  datastore master nodes: 127.0.0.1:19001
  datastore standby nodes: none`

**Comment by davidbarratt** [comment]
Running
```
sudo microk8s refresh-certs
```
seems to have _helped_ ... maybe?

**Comment by alifiroozi80** [comment]
For me Ubuntu 20.04.4 LTS (Focal Fossa), this two commands save my day:

`sudo microk8s.refresh-certs --cert ca.crt`

and

`sudo microk8s.refresh-certs --cert server.crt`

**Comment by arshpreetsingh** [comment]
I am getting following while enabling istio.

$ microk8s enable istio


Enabling DNS
Applying manifest
error: Get "https://127.0.0.1:16443/openapi/v2?timeout=32s": dial tcp 127.0.0.1:16443: connect: connection refused - error from a previous attempt: unexpected EOF

**Comment by mitul3737** [comment]
I have solved this issue. I have uninstalled microK8s and then installed it.
The problem got solved !!!!

**Comment by JJMoon** [comment]
I got this error when I input some type in cert-manager. The cluster was rebooting (or resetting ? anyway) repeatedly.
After I fix the typos, the rebooting was not ended.
The commands of @alifiroozi80 makes the cluster stop the rebooting.
 I assume that @alifiroozi80 's refresh command reset the certificates.

But it reset my namespace context also. So I set the context again. FYI
```
microk8s cubectl config view # It shows current namespace context
microk8s cubectl config set-context microk8s --namespace=your-namespace
```
